ormar/queryset/queryset.py

```diff
@@ -1055,6 +1056,38 @@ async def all(self, *args: Any, **kwargs: Any) -> List["T"]:  # noqa: A003
 
         return result_rows
 
+    async def iterator(  # noqa: A003
+        self,
+        *args: Any,
+        **kwargs: Any,
+    ) -> AsyncGenerator["T", None]:
+        """
+        Return async iterable generator for all rows from a database for given model.
+
+        Passing args and/or kwargs is a shortcut and equals to calling
+        `filter(*args, **kwargs).iterator()`.
+
+        If there are no rows meeting the criteria an empty async generator is returned.
+
+        :param kwargs: fields names and proper value types
+        :type kwargs: Any
+        :return: asynchronous iterable generator of returned models
+        :rtype: AsyncGenerator[Model]
+        """
+
+        if kwargs or args:
+            async for result in self.filter(*args, **kwargs).iterator():
+                yield result
+            return
+
+        expr = self.build_select_expression()
+        async for row in self.database.iterate(query=expr):
+            result_row = self._process_query_result_rows([row])
```

Comment:
Now that I think about this - it won't work for models with relations in this way.

Since in the case of a join one model is spread i.e. through three database rows (so 3 child models) if you iterate like this you would get three times the same parent with a different child each time. 

We had a similar problem with limit and offset queries where you should get one full parent model if you limit 1 and not only the first database row which is not the entire parent model.

Reply:
Hello again @collerek .
Thank you for reading and reviewing.

Yes, you are absolutely right;
But we still have this problem in limit and offset:
For example, when you want to limit the results with limit, if you do not True `limit_raw_sql`, we get an error,
And if we set its value to True, the restriction applies to the number of child models, not to the parent!

The same thing happens in the iterator,
which is repeated for the number of child models of the parent model;
But it works perfectly well on unrelated models.

Create an issue for the problem I mentioned?
What solution do you suggest to solve this problem?

Reply:
> Now that I think about this - it won't work for models with relations in this way.
> 
> Since in the case of a join one model is spread i.e. through three database rows (so 3 child models) if you iterate like this you would get three times the same parent with a different child each time.
> 
> We had a similar problem with limit and offset queries where you should get one full parent model if you limit 1 and not only the first database row which is not the entire parent model.

Hello again @collerek 
Thank you for reading and reviewing.

Yes, you are absolutely right;
But we still have this problem in limit and offset:
For example, when you want to limit the results with limit, if you do not True `limit_raw_sql`, we get an error,
And if we set its value to True, the restriction applies to the number of child models, not to the parent!

The same thing happens in the iterator,
which is repeated for the number of child models of the parent model;
But it works perfectly well on unrelated models.

Create an issue for the problem I mentioned?
What solution do you suggest to solve this problem?

Reply:
What error do you get? It should work fine, so if there is some error it is probably a bug. Yes, please create an issue for it if you found something. Thanks 🙂

Reply:
> What error do you get? It should work fine, so if there is some error it is probably a bug. Yes, please create an issue for it if you found something. Thanks 🙂

I explained it in issue #704

Reply:
Hi again @collerek 
Thank you for fixing issue #704 

I reviewed the relevant codes again and finally came to the following conclusion.
Thank you for your help and suggestions.

We have two choices for the iterator:
1. Or get all the rows together with `fetch_all` and then write the model build as an async generator to be able to join.
2. The second approach is that if we want to use the `iterate` method of the `databases` package, because one line is received each time, we will not be able to use commands such as `select_related` and etc. when using this method.

Reply:
Hi, sorry for taking a long time to review.

I belive we can make it work with select related (yield from db iterator until pk of the main model changes and only then yield from ormar).

That will require some additional work so if you need this urgently you can modify the QuerySet used and simply add your method to queryset and use it with the current version. Modifing the queryset is described here: https://collerek.github.io/ormar/models/#overwriting-the-default-queryset

So basically you can do:

```python
class IterableQuerySet(QuerySet):

    async def iterator(  # noqa: A003
        self,
        *args: Any,
        **kwargs: Any,
    ) -> AsyncGenerator["T", None]:
        """
        Return async iterable generator for all rows from a database for given model.
        Passing args and/or kwargs is a shortcut and equals to calling
        `filter(*args, **kwargs).iterator()`.
        If there are no rows meeting the criteria an empty async generator is returned.
        :param kwargs: fields names and proper value types
        :type kwargs: Any
        :return: asynchronous iterable generator of returned models
        :rtype: AsyncGenerator[Model]
        """

        if kwargs or args:
            async for result in self.filter(*args, **kwargs).iterator():
                yield result
            return

        expr = self.build_select_expression()
        async for row in self.database.iterate(query=expr):
            result_row = self._process_query_result_rows([row])
            if self._prefetch_related and result_row:
                result_row = await self._prefetch_related_models(result_row, [row])

            yield result_row[0]


class YourModel(ormar.Model):

    class Meta(ormar.ModelMeta):
        ...
        queryset_class = IterableQuerySet
    ...

```

Reply:
thank you for answering 
I did it by what you suggested and I am relieved to say It worked! I hadn't thought of this
So sorry to bother you and take so much of your time 

I tried so many times and all I got were errors! Therefore sorry that I committed  so many times 

Thanks for your help